src/statsmodels/formula.jl

```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+            haskey(eterm_cols, et_and_nr) || 
+                setindex!(eterm_cols,
+                          modelmat_cols(et_and_nr[1], mf, non_redundant=et_and_nr[2]),
+                          et_and_nr)
+            push!(term_cols, eterm_cols[et_and_nr])
+        end
+        push!(mm_cols, term_cols)
     end
-    ModelMatrix{Float64}(hcat([expandcols(t) for t in aa]...), asgn)
+
+    ## TODO: could this be made more efficient by
+    ## first computing mm_col_term_nums, initializing mm, and directly indexing?
+    mm = hcat([expandcols(tc) for tc in mm_cols]...)
+    mm_col_term_nums = vcat([fill(i_term, nc(tc)) for (i_term,tc) in enumerate(mm_cols)]...)
+
+    ModelMatrix{Float64}(mm, mm_col_term_nums)
+
 end
 
+
+
 termnames(term::Symbol, col) = [string(term)]
-function termnames(term::Symbol, col::PooledDataArray)
-    levs = levels(col)
-    [string(term, " - ", levs[i]) for i in 2:length(levs)]
+function termnames(term::Symbol, mf::ModelFrame; non_redundant::Bool = false)
+    if haskey(mf.contrasts, term)
+        termnames(term, mf.df[term], 
+             non_redundant ? promote_contrast(mf.contrasts[term]) : mf.contrasts[term])
+    else
+        termnames(term, mf.df[term])
+    end
 end
+termnames(term::Symbol, col::Any, contrast::ContrastsMatrix) =
+    ["$term: $name" for name in contrast.termnames]
 
-function coefnames(mf::ModelFrame)
-    if mf.terms.intercept
-        vnames = UTF8String["(Intercept)"]
+
+function expandtermnames(term::Vector)
+    if length(term) == 1
+        return term[1]
     else
-        vnames = UTF8String[]
-    end
-    # Need to only include active levels
-    for term in mf.terms.terms
-        if isa(term, Expr)
-            if term.head == :call && term.args[1] == :|
-                continue                # skip random-effects terms
-            elseif term.head == :call && term.args[1] == :&
-                ## for an interaction term, combine term names pairwise,
-                ## starting with rightmost terms
-                append!(vnames,
-                        foldr((a,b) ->
-                              vec([string(lev1, " & ", lev2) for
+        return foldr((a,b) -> vec([string(lev1, " & ", lev2) for
                                    lev1 in a,
                                    lev2 in b]),
-                              map(x -> termnames(x, mf.df[x]), term.args[2:end])))
-            else
-                error("unrecognized term $term")
-            end
-        else
-            append!(vnames, termnames(term, mf.df[term]))
+                     term)
+    end
+end
+
+function coefnames(mf::ModelFrame)
+    ## strategy mirrors ModelMatrx constructor:
+    eterm_names = Dict{Tuple{Symbol,Bool}, Vector{String}}()
+    term_names = Vector{Vector{String}}[]
+    mf.terms.intercept && push!(term_names, Vector[String["(Intercept)"]])
+
+    factors = round(Bool, mf.terms.factors)
+
+    for (i_term, term) in enumerate(mf.terms.terms)
```

Comment:
There's not ambiguity in principle, but I (at least) find it helps in reading the code to know when we're indexing terms vs. evalterms or something else. But perhaps that's just lingering PTST from all the single-letter index/variable names that used to be in these functions...